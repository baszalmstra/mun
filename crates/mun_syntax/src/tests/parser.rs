use crate::SourceFile;

fn snapshot_test(text: &str) {
    let text = text.trim().replace("\n    ", "\n");
    let file = SourceFile::parse(&text);
    insta::assert_snapshot!(insta::_macro_support::AutoName, file.debug_dump(), &text);
}

#[test]
fn empty() {
    snapshot_test(r#""#);
}

#[test]
fn function() {
    snapshot_test(
        r#"
    // Source file comment

    // Comment that belongs to the function
    fn a() {}
    fn b(value:number) {}
    export fn c():never {}
    fn b(value:number):number {}"#,
    );
}

#[test]
fn block() {
    snapshot_test(
        r#"
    fn foo() {
        let a;
        let b:i32;
        let c:string;
    }"#,
    );
}

#[test]
fn literals() {
    snapshot_test(
        r#"
    fn foo() {
        let a = true;
        let b = false;
        let c = 1;
        let d = 1.12;
        let e = "Hello, world!"
    }
    "#,
    );
}

#[test]
fn unary_expr() {
    snapshot_test(
        r#"
    fn foo() {
        let a = --3;
        let b = !!true;
    }
    "#,
    )
}

#[test]
fn binary_expr() {
    snapshot_test(
        r#"
    fn foo() {
        let a = 3+4*5
        let b = 3*4+10/2
    }
    "#,
    )
}

#[test]
fn expression_statement() {
    snapshot_test(
        r#"
    fn foo() {
        let a = "hello"
        let b = "world"
        let c
        b = "Hello, world!"
        !-5+2*(a+b);
        -3
    }
    "#,
    )
}

#[test]
fn function_calls() {
    snapshot_test(
        r#"
    fn bar(i:number) { }
    fn foo(i:number) {
      bar(i+1)
    }
    "#,
    )
}

#[test]
fn patterns() {
    snapshot_test(
        r#"
    fn main(_:number) {
       let a = 0;
       let _ = a;
    }
    "#,
    )
}

#[test]
fn compare_operands() {
    snapshot_test(
        r#"
    fn main() {
        let _ = a==b;
        let _ = a == b;
        let _ = a != b;
        let _ = a < b;
        let _ = a > b;
        let _ = a <= b;
        let _ = a >= b;
    }
    "#,
    )
}

#[test]
fn if_expr() {
    snapshot_test(
        r#"
    fn bar() {
        if true {};
        if true {} else {};
        if true {} else if false {} else {};
        if {true} {} else {}
    }
    "#,
    );
}

#[test]
fn block_expr() {
    snapshot_test(
        r#"
    fn bar() {
        {3}
    }
    "#,
    );
}

#[test]
fn return_expr() {
    snapshot_test(
        r#"
    fn foo() {
        return;
        return 50;
    }
    "#,
    )
}

#[test]
fn loop_expr() {
    snapshot_test(
        r#"
    fn foo() {
        loop {}
    }"#,
    )
}

#[test]
fn break_expr() {
    snapshot_test(
        r#"
    fn foo() {
        break;
        if break { 3; }
        if break 4 { 3; }
    }
    "#,
    )
}

#[test]
fn while_expr() {
    snapshot_test(
        r#"
    fn foo() {
        while true {};
        while { true } {};
    }
    "#,
    )
}

#[test]
fn range_expr() {
    snapshot_test(
        r#"
    fn foo() {
        1..2;
        3..;
        ..4;
        ..;
        5..=6;
        ..=7;
        3..=; // Error: an inclusive range must have an end expression
        ..=;  // Error: an inclusive range must have an end expression
    }
    "#,
    )
}

#[test]
fn postfix_range() {
    snapshot_test(
        r#"
    fn foo() {
        let x = 1..;
        if 1.. {};
    }
    "#,
    )
}
