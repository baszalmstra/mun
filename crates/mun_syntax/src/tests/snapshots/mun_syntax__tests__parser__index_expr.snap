---
source: crates/mun_syntax/src/tests/parser.rs
expression: "fn main() {\n    let a = [1,2,3,4]\n    let b = a[0];\n    let c = a[b];\n    a[0] = c;\n    let a = { [3,4,5] }[1];\n}"
---
SOURCE_FILE@0..113
  FUNCTION_DEF@0..113
    FN_KW@0..2 "fn"
    WHITESPACE@2..3 " "
    NAME@3..7
      IDENT@3..7 "main"
    PARAM_LIST@7..9
      L_PAREN@7..8 "("
      R_PAREN@8..9 ")"
    WHITESPACE@9..10 " "
    BLOCK_EXPR@10..113
      L_CURLY@10..11 "{"
      WHITESPACE@11..16 "\n    "
      LET_STMT@16..33
        LET_KW@16..19 "let"
        WHITESPACE@19..20 " "
        BIND_PAT@20..21
          NAME@20..21
            IDENT@20..21 "a"
        WHITESPACE@21..22 " "
        EQ@22..23 "="
        WHITESPACE@23..24 " "
        ARRAY_EXPR@24..33
          L_BRACKET@24..25 "["
          LITERAL@25..26
            INT_NUMBER@25..26 "1"
          COMMA@26..27 ","
          LITERAL@27..28
            INT_NUMBER@27..28 "2"
          COMMA@28..29 ","
          LITERAL@29..30
            INT_NUMBER@29..30 "3"
          COMMA@30..31 ","
          LITERAL@31..32
            INT_NUMBER@31..32 "4"
          R_BRACKET@32..33 "]"
      WHITESPACE@33..38 "\n    "
      LET_STMT@38..51
        LET_KW@38..41 "let"
        WHITESPACE@41..42 " "
        BIND_PAT@42..43
          NAME@42..43
            IDENT@42..43 "b"
        WHITESPACE@43..44 " "
        EQ@44..45 "="
        WHITESPACE@45..46 " "
        INDEX_EXPR@46..50
          PATH_EXPR@46..47
            PATH@46..47
              PATH_SEGMENT@46..47
                NAME_REF@46..47
                  IDENT@46..47 "a"
          L_BRACKET@47..48 "["
          LITERAL@48..49
            INT_NUMBER@48..49 "0"
          R_BRACKET@49..50 "]"
        SEMI@50..51 ";"
      WHITESPACE@51..56 "\n    "
      LET_STMT@56..69
        LET_KW@56..59 "let"
        WHITESPACE@59..60 " "
        BIND_PAT@60..61
          NAME@60..61
            IDENT@60..61 "c"
        WHITESPACE@61..62 " "
        EQ@62..63 "="
        WHITESPACE@63..64 " "
        INDEX_EXPR@64..68
          PATH_EXPR@64..65
            PATH@64..65
              PATH_SEGMENT@64..65
                NAME_REF@64..65
                  IDENT@64..65 "a"
          L_BRACKET@65..66 "["
          PATH_EXPR@66..67
            PATH@66..67
              PATH_SEGMENT@66..67
                NAME_REF@66..67
                  IDENT@66..67 "b"
          R_BRACKET@67..68 "]"
        SEMI@68..69 ";"
      WHITESPACE@69..74 "\n    "
      EXPR_STMT@74..83
        BIN_EXPR@74..82
          INDEX_EXPR@74..78
            PATH_EXPR@74..75
              PATH@74..75
                PATH_SEGMENT@74..75
                  NAME_REF@74..75
                    IDENT@74..75 "a"
            L_BRACKET@75..76 "["
            LITERAL@76..77
              INT_NUMBER@76..77 "0"
            R_BRACKET@77..78 "]"
          WHITESPACE@78..79 " "
          EQ@79..80 "="
          WHITESPACE@80..81 " "
          PATH_EXPR@81..82
            PATH@81..82
              PATH_SEGMENT@81..82
                NAME_REF@81..82
                  IDENT@81..82 "c"
        SEMI@82..83 ";"
      WHITESPACE@83..88 "\n    "
      LET_STMT@88..107
        LET_KW@88..91 "let"
        WHITESPACE@91..92 " "
        BIND_PAT@92..93
          NAME@92..93
            IDENT@92..93 "a"
        WHITESPACE@93..94 " "
        EQ@94..95 "="
        WHITESPACE@95..96 " "
        BLOCK_EXPR@96..107
          L_CURLY@96..97 "{"
          WHITESPACE@97..98 " "
          ARRAY_EXPR@98..105
            L_BRACKET@98..99 "["
            LITERAL@99..100
              INT_NUMBER@99..100 "3"
            COMMA@100..101 ","
            LITERAL@101..102
              INT_NUMBER@101..102 "4"
            COMMA@102..103 ","
            LITERAL@103..104
              INT_NUMBER@103..104 "5"
            R_BRACKET@104..105 "]"
          WHITESPACE@105..106 " "
          R_CURLY@106..107 "}"
      EXPR_STMT@107..111
        ARRAY_EXPR@107..110
          L_BRACKET@107..108 "["
          LITERAL@108..109
            INT_NUMBER@108..109 "1"
          R_BRACKET@109..110 "]"
        SEMI@110..111 ";"
      WHITESPACE@111..112 "\n"
      R_CURLY@112..113 "}"

